src/rdf4cpp/rdf/storage/node/reference_node_storage/ReferenceNodeStorageBackend.cpp

```diff
@@ -1,144 +1,259 @@
-                assert(inserted_successfully);
-                found = std::move(found2);
-                storage.id2data.emplace(id, found->first.get());
-                return id;
-            } else {
-                unique_lock.unlock();
-                return found->second;
-            }
-        } else {
-            return {};
+template<bool create_if_not_present, typename Storage, typename NextIDFunc = void *>
+    requires (!create_if_not_present || std::is_nothrow_invocable_r_v<identifier::NodeID, NextIDFunc>)
+static identifier::NodeID lookup_or_insert_impl(typename Storage::BackendView const &view,
+                                                Storage &storage,
+                                                NextIDFunc next_id_func = nullptr) noexcept {
+
+    {
+        std::shared_lock lock{storage.mutex};
+        if (auto const it = storage.data2id.find(view); it != storage.data2id.end()) {
+            return it->second;
         }
+    }
+
+    if constexpr (!create_if_not_present) {
+        return identifier::NodeID{};
     } else {
-        shared_lock.unlock();
-        return found->second;
+        std::unique_lock lock{storage.mutex};
+
+        // check again, might have changed between unlocking of shared_lock and locking of unique_lock
+        if (auto const it = storage.data2id.find(view); it != storage.data2id.end()) {
+            return it->second;
+        }
+
+        identifier::NodeID const next_id = next_id_func();
+        auto const [it, inserted] = storage.id2data.emplace(next_id, std::make_unique<typename Storage::Backend>(view));
+        assert(inserted);
+        storage.data2id.emplace(it->second.get(), next_id);
+
+        return next_id;
     }
 }
 
 identifier::NodeID ReferenceNodeStorageBackend::find_or_make_id(view::LiteralBackendView const &view) noexcept {
-    return lookup_or_insert_impl<LiteralBackend, true>(
-            view, literal_storage_,
-            [this]([[maybe_unused]] view::LiteralBackendView const &literal_view) {
-                return identifier::NodeID{next_literal_id++,
-                                          identifier::iri_node_id_to_literal_type(literal_view.datatype_id)};
+    return view.visit(
+            [this](view::LexicalFormLiteralBackendView const &lexical) noexcept {
+                auto const next_id_func = [this, datatype = lexical.datatype_id]() noexcept {
+                    return identifier::NodeID{next_literal_id++, identifier::iri_node_id_to_literal_type(datatype)};
+                };
+
+                return lookup_or_insert_impl<true>(lexical, fallback_literal_storage_, next_id_func);
+            },
+            [this](view::ValueLiteralBackendView const &any) noexcept {
+                auto const next_id_func = [this, datatype = any.datatype]() noexcept {
+                    return identifier::NodeID{next_literal_id++, datatype};
```

Comment:
We need a next_literal_id per LiteralType::underlying_type. So for 2^6 types. For the types that consume more than 42-bits we should probably define a bit-depth threshold up to what they are inlined and reserve the remaining IDs for other values. 

Also, we should throw an exception or terminate if ids are exceeded.  

Addressing this might be something to create an issue and adress in a subsequent PR. Makes reviewing it much easier.

Reply:
Already prepared the PR #148 by branching off this branch. You can review it when this is merged